### DISCUSSION TITLE
Alpha issues

### DISCUSSION THREAD

**Discussion by vurt72** [BadDogSkyrim/PyNifly#161] [General]
Tried to compare where the alpha does look right, but i can't say i understand what's different with those meshes. 
Also meshes with perfect looking alpha at export with Nifly looks bad again with terrible clipping when importing them again.

https://www.youtube.com/watch?v=WKnJiF1m_yQ&ab_channel=Vurt_NexusMods

Makes it very hard to work with 2D planes, which to 95% is what i do... 
Import any grass mesh from Skyrim SE and you will see the issue, some does work though (forestgrass02 for example has no clipping).

**Comment by mindflvx** [comment]
Which one for example is not being displayed correctly? I can check too.

**Comment by mindflvx** [comment]
I see the issue, it's due to the VERTEX_ALPHA attribute in the shader network setting the minimum transparency. Not entirely sure how this is supposed to be work at Blender side but for now you can work around the issue by piping alpha (from color map) directly into alpha input of Principled BSDF (or simply removing map range node input from the map range going into alpha).

**Reply by vurt72** [reply]
Thank you! I will try that.

**Comment by BadDogSkyrim** [comment]
Huh. That map node was intended to give the same visual effect in Blender that you saw in game. Wasn't supposed to change the output value. I'll look at that.

**Reply by BadDogSkyrim** [reply]
Not seeing an issue. The youtube you linked to doesn't say what mesh you were using. I would expect it to be TundraGrassObj03, but when I open that nif I get different textures. Have you been playing around?

TundraGrassObj04 does play with the alpha, presumably to set the amount of waving, but I can export it and get exactly what I got on import, so I don't see a bug. 

It would be better not to treat the alpha channel as a texture alpha for that nif, but I don't see anything signaling that it's not a normal alpha. It's the usual BSLightingShaderProperty using the default shader. So I'm pretty much stuck doing what I do now.

**Comment by vurt72** [comment]
its my own texture atlas i tried there.
but 
tundragrass01.nif has this exact thing
blackreachgrassobj01.nif as well, for that mesh no texture at all shows up.
forestgrass01.nif does not have it.
i'm importing from Skyrim SE

the fix is easy enough though, just run alpha from
the texture to principled bsdf alpha

yeah it uses vertex alpha painting for wind so maybe it's something with that.

**Comment by mindflvx** [comment]
Agree, I don't think vertex alpha should be used for viewport transparency since as far as I'm aware it's only used for setting up wind. Visualization is not really a problem since you can always go to vertex paint mode with the added benefit of being able to modify the colors while at it.

I tested with unmodified vanilla grass object beachgrassobj01.nif with vanilla textures.

**Comment by mindflvx** [comment]
But there is indeed a problem of how to detect the grass objects. As a matter of fact Nifskope seemingly has the same problem since these objects are displayed showing essentially the same issue with viewport visualziation. Alas, the only difference in regards to shaders that I can see is that (vanilla) grass uses very specific alpha testing (source one, dest one) properties that I'm not aware any other asset uses. But modded grass objects can probably use anything.

**Reply by BadDogSkyrim** [reply]
The problem is that vertex colors and vertex alpha can be used very differently by different nifs, and usually vertex alpha sets the threshold between opaque and transparent. Usually it's the shader type that decides how vertex alpha is used, but with these grass models apparently it's just the game that handles them differently. I can have a look at those alpha flags but I'm inclined to say just turn of alpha in the viewport and deal with it. But I will look.

**Reply by mindflvx** [reply]
Thanks. I'm guessing there's no robust way to tell these apart from other stuff, seeing as the alpha flags are very likely not trustworthy.

You're right, as far as I remember the game uses hardcoded shaders for the instanced grass (i.e. the type of grass we're looking at) and while one can tweak the shaders as much as you want not much if anything makes any difference in the rendering. Apparently leaf animation too is employed by the shader without needing to specify it in the shader flags.

**Reply by BadDogSkyrim** [reply]
The importer creates a shader that runs the alpha input from a "Attribute" node that reads the vertex alpha value. In the map node where that value is plugged in, disconnect the attribute node and set the value to 0 (or 1, whichever works). That won't screw up export but it will look fine in blender.
